### ISSUE TITLE
Error while executing train.py

### ISSUE THREAD

**Issue by Mrugankakarte** [LIBOL/ODL#1]
Hello Sir,

I am getting an error  `run() got an unexpected keyword argument 'hedge'` while executing the train.py

Can you please let me know what might be cause of it?

Thanks & Regards
Mrugank Akarte

**Comment by doyensahoo** [comment]
Hi Mrugank, Sorry for the late response.

You may look at the requirements and installation section:
To install HBP, you need to replace the Keras's keras/engine/training.py file with our modified training.py. this doesn't affect normal projects that don't use HBP. Note that as the current HBP implementation only supports Keras 1.

This should address your issue.

**Comment by Mrugankakarte** [comment]
Thanks for your response. I am using keras 2.1.9, So that might be the
issue. Actually I am trying to implement your work in R, I was able to get
rid of the error but the loss for each output layer (out1, out2,..)was
same. Does the keras backpropgation updates the weights the way you have
mentioned in paper?

Thanks
Mrugank Akarte

**Comment by doyensahoo** [comment]
The Keras version we used did not. That's why we modified the back-end training file, in order to implement our proposed Hedge-Backpropagation scheme.

**Comment by Mrugankakarte** [comment]
Sir do you have any plans on updating the file for current keras version?

**Comment by doyensahoo** [comment]
Hi Mrugank, We'll keep this issue open and try to do it soon in the future. 
But it maybe difficult for us to do it in next 3-4 weeks, due to some deadlines. 

Maybe you can try to use the older version of Keras, and see if it works for you?

**Comment by Mrugankakarte** [comment]
Thanks for the update. I will try with the older version of keras for implementation.

**Comment by Mrugankakarte** [comment]
Sir,

Can you please explain me the update step? 

My interpretation is as follows. Please correct me.
Lets assume a 3 hidden layer network. so NN will look like ....
* input size (m, 28) ,  m=number of data points and no of features per data point are 28
* hidden units per layer  = 10
* output units = 1 (since we only have two classes 0 and 1)
* alpha = [0.25, 0.25, 0.25, 0.25]  (alpha0, alpha1, alpha2, alpha3)
* lr = learning rate = 0.01
```
input -> h1 = g(z1 = X.w1 ) -> h2 = g(z2 = h1.w2) -> h3 = g(z3 = h2.w3)
  |            |                      |                         |
  f0          f1                      f2                        f3

```
classifiers are now:
```
f0 = softmax(o0.X)
f1 = softmax(o1.h1)
f2 = softmax(o2.h2)
f3 = softmax(o3.h3)
```
dimensions of matrices 'w' and 'o' will be as follows:
```
X = (m, 28)
w0 = does not exists
w1 = (28, 10)
w2 = (10, 10)
w3 = (10, 10)

o0 = (28, 1)
o1 = (10, 1)
o2 = (10, 1)
o3 = (10, 1)
```
loss 
```
loss0 = crossentropy(f0, Y)
loss1 = crossentropy(f1, Y)
loss2 = crossentropy(f2, Y)
loss3 = crossentropy(f3, Y)

total_loss = alpha0 * loss0 + alpha1 * loss1 + alpha2 * loss2 + alpha3 * loss3
```
updating weights 
```
alpha0 = alpha0 * beta^(loss0)
alpha1 = alpha1 * beta^(loss1)
alpha2 = alpha2 * beta^(loss2)
alpha3 = alpha3 * beta^(loss3)

o0  = o0 - alpha0 * lr * gradient of (o0 with respect to loss0)
o1  = o1 - alpha1 * lr * gradient of (o1 with respect to loss1)
o2  = o2 - alpha2 * lr * gradient of (o2 with respect to loss2)
o3  = o3 - alpha3 * lr * gradient of (o3 with respect to loss3)

w1 = w1 - lr * alpha1 * SUM(gradient of (w1 with respect to loss1), gradient of (w1 with respect to loss2), gradient of (w1 with respect to loss3))
w2 = w2 - lr * alpha2 * SUM(gradient of (w2 with respect to loss2), gradient of (w2 with respect to loss3))
w3 = w3 - lr * alpha3 * SUM(gradient of (w3 with respect to loss3))
```

~~I do not understand how can we sum dw1, dw2, dw3 as their dimensions are different.~~ Is my interpretation correct or am I missing somrthing. Can you please explain me this weight update part?

Thanking you

Mrugank Akarte

**Comment by doyensahoo** [comment]
Your interpretation appears to be correct. Are there any issues you are still facing?

**Comment by Mrugankakarte** [comment]
Just the implementation issue sir. Since this is a multi-output model, frameworks like keras do update the model weights as
```
w1 = w1 - lr  * SUM(gradient of (w1 with respect to loss1), gradient of (w1 with respect to loss2), gradient of (w1 with respect to loss3))
w2 = w2 - lr  * SUM(gradient of (w2 with respect to loss2), gradient of (w2 with respect to loss3))
w3 = w3 - lr  * SUM(gradient of (w3 with respect to loss3))
```
so the only question remains is how to multiply the gradients with `alpha` before updating. The values of `alpha` are being updated using callback. But how are these `alpha` values being multiplied to gradient? 
 Is it through the `loss_weight` parameter? Does the modified `training.py` takes care of this problem or is it modified just to add the `hedge` parameter which acts as a switch between original and hedge-backpropagation? I tried comparing the original `training.py` with the modified one, but couldn't understand the exact difference. Can you please share what part is modified and how it affects the execution?

**Comment by phquang** [comment]
Hi, you can look at our training.py file, line 678-684 where the total loss is computed as a dot product between the hedge weight vector and the loss vector (self.holder is the hedge weights , self.metrics_tensors is the losses).
So the remaining step is to set the hedge weights before each computation in training and testing, which you can look at line 922 (training), line 1021 (testing).

**Comment by Mrugankakarte** [comment]
Thank you so much Sir for pointing that out. So the idea is to compute `total cost = hedge weights * losses` and while back-propagating, these hedge weights will be automatically carried back to update the weights of model. Am I correct?

**Comment by phquang** [comment]
Yes, you got it right.

**Comment by Mrugankakarte** [comment]
Ok. Thanks a lot Sir
